library/std/src/sys/uefi/pipe.rs

```diff
@@ -0,0 +1,188 @@
+        self.handle
+    }
+
+    pub fn read(&self, buf: &mut [u8]) -> io::Result<usize> {
+        let protocol = common::open_protocol(self.handle, uefi_pipe_protocol::PROTOCOL_GUID)?;
+        unsafe { uefi_pipe_protocol::Protocol::read(protocol.as_ptr(), buf) }
+    }
+
+    pub(crate) fn read_to_end(&self, buf: &mut Vec<u8>) -> io::Result<usize> {
+        let protocol = common::open_protocol(self.handle, uefi_pipe_protocol::PROTOCOL_GUID)?;
+        unsafe { uefi_pipe_protocol::Protocol::read_to_end(protocol.as_ptr(), buf) }
+    }
+
+    pub fn read_vectored(&self, bufs: &mut [IoSliceMut<'_>]) -> io::Result<usize> {
+        crate::io::default_read_vectored(|buf| self.read(buf), bufs)
+    }
+
+    #[inline]
+    pub fn is_read_vectored(&self) -> bool {
+        false
+    }
+
+    pub fn write(&self, buf: &[u8]) -> io::Result<usize> {
+        let protocol = common::open_protocol(self.handle, uefi_pipe_protocol::PROTOCOL_GUID)?;
+        unsafe { uefi_pipe_protocol::Protocol::write(protocol.as_ptr(), buf) }
+    }
+
+    pub fn write_vectored(&self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
+        crate::io::default_write_vectored(|buf| self.write(buf), bufs)
+    }
+
+    #[inline]
+    pub fn is_write_vectored(&self) -> bool {
+        false
+    }
+
+    pub fn diverge(&self) -> ! {
+        unimplemented!()
+    }
+}
+
+pub fn read2(p1: AnonPipe, v1: &mut Vec<u8>, p2: AnonPipe, v2: &mut Vec<u8>) -> io::Result<()> {
+    p1.read_to_end(v1)?;
+    p2.read_to_end(v2)?;
+    Ok(())
+}
+
+pub(crate) mod uefi_pipe_protocol {
+    use crate::collections::VecDeque;
+    use crate::io;
+    use crate::sys::uefi::common;
+    use io::{Read, Write};
+    use r_efi::efi::Guid;
+
+    pub(crate) const PROTOCOL_GUID: Guid = Guid::from_fields(
+        0x3c4acb49,
+        0xfb4c,
+        0x45fb,
+        0x93,
+        0xe4,
+        &[0x63, 0x5d, 0x71, 0x48, 0x4c, 0x0f],
+    );
+
+    // Maybe the internal data needs to be wrapped in a Mutex?
+    #[repr(C)]
+    #[derive(Debug)]
+    pub(crate) struct Pipedata {
+        data: VecDeque<u8>,
+    }
+
+    impl Pipedata {
+        #[inline]
+        pub(crate) fn with_capacity(capacity: usize) -> Pipedata {
+            Pipedata { data: VecDeque::with_capacity(capacity) }
+        }
+
+        #[inline]
+        pub(crate) unsafe fn read(data: *mut Pipedata, buf: &mut [u8]) -> io::Result<usize> {
+            unsafe { (*data).data.read(buf) }
+        }
+
+        #[inline]
+        pub(crate) unsafe fn read_to_end(
+            data: *mut Pipedata,
+            buf: &mut Vec<u8>,
+        ) -> io::Result<usize> {
+            unsafe { (*data).data.read_to_end(buf) }
+        }
+
+        #[inline]
+        pub(crate) unsafe fn write(data: *mut Pipedata, buf: &[u8]) -> io::Result<usize> {
+            unsafe { (*data).data.write(buf) }
+        }
+    }
+
+    #[repr(C)]
+    pub(crate) struct Protocol {
+        data: *mut Pipedata,
```

Comment:
The below methods should probably be function pointers here, so that if, for example, `VecDeque` changes, this protocol doesn't break.

Reply:
Since this protocol was internal to std, I used associated functions (`pub(crate)` and thus kinda make using the protocol not possible from outside). I don't mind adding function pointers, but should I (since it is internal)?

Reply:
Since this protocol is shared by different processes, they are not guaranteed to be using the same Rust versions, so `VecDeque` might use a different ABI or implementation.